src/index.js

```diff
@@ -0,0 +1,725 @@
+            tosend.add(peer)
+          }
+        })
+
+        // Gossipsub peers handling
+        let meshPeers = this.mesh.get(topic)
+        if (!meshPeers) {
+          // We are not in the mesh for topic, use fanout peers
+          meshPeers = this.fanout.get(topic)
+          if (!meshPeers) {
+            // If we are not in the fanout, then pick any peers in topic
+            const peers = this._getPeers(topic, constants.GossipSubD)
+
+            if (peers.size > 0) {
+              meshPeers = peers
+              this.fanout.set(topic, peers)
+            } else {
+              meshPeers = []
+            }
+          }
+          // Store the latest publishing time
+          this.lastpub.set(topic, this._now())
+        }
+
+        meshPeers.forEach((peer) => {
+          tosend.add(peer)
+        })
+      })
+      // Publish messages to peers
+      tosend.forEach((peer) => {
+        if (peer.info.id.toB58String() === msgObj.from) {
+          return
+        }
+        this._sendRpc(peer, { msgs: [msgObj] })
+      })
+    })
+  }
+
+  /**
+   * Sends a GRAFT message to a peer
+   *
+   * @param {Peer} peer
+   * @param {String} topic
+   * @returns {void}
+   */
+  _sendGraft (peer, topic) {
+    const graft = [{
+      topicID: topic
+    }]
+
+    const out = this._rpcWithControl(null, null, null, graft, null)
+    this._sendRpc(peer, out)
+  }
+
+  /**
+   * Sends a PRUNE message to a peer
+   *
+   * @param {Peer} peer
+   * @param {String} topic
+   * @returns {void}
+   */
+  _sendPrune (peer, topic) {
+    const prune = [{
+      topicID: topic
+    }]
+
+    const out = this._rpcWithControl(null, null, null, null, prune)
+    this._sendRpc(peer, out)
+  }
+
+  _sendRpc (peer, outRpc) {
+    if (!peer || !peer.isWritable) {
+      return
+    }
+
+    // piggyback control message retries
+    const ctrl = this.control.get(peer)
+    if (ctrl) {
+      this.piggybackControl(peer, outRpc, ctrl)
+      this.control.delete(peer)
+    }
+
+    // piggyback gossip
+    const ihave = this.gossip.get(peer)
+    if (ihave) {
+      this.piggybackGossip(peer, outRpc, ihave)
+      this.gossip.delete(peer)
+    }
+
+    peer.write(rpc.RPC.encode(outRpc))
+  }
+
+  _piggybackControl (peer, outRpc, ctrl) {
+    const hasPeerInTopic = (topicID) => {
+      const meshPeers = this.mesh.get(topicID)
+      return meshPeers && meshPeers.has(peer)
+    }
+    const tograft = (ctrl.graft || []).filter(({ topicID }) => hasPeerInTopic(topicID))
+    const toprune = (ctrl.prune || []).filter(({ topicID }) => hasPeerInTopic(topicID))
+
+    if (!tograft.length && !toprune.length) {
+      return
+    }
+
+    outRpc.control.graft = outRpc.control.graft.concat(tograft)
+    outRpc.control.prune = outRpc.control.graft.concat(toprune)
+  }
+
+  _piggybackGossip (peer, outRpc, ihave) {
+    outRpc.control.ihave = ihave
+  }
+
+  /**
+   * Maintains the mesh and fanout maps in gossipsub.
+   *
+   * @returns {void}
+   */
+  _heartbeat () {
```

Comment:
I would like to have the `heartbeat` logic in a separate file, a class that receives a reference to the properties it needs from here, and has the start and stop functions, which are called from here.